gradient/commands/datasets.py

```diff
@@ -557,24 +560,114 @@ def update_status():
         with requests.Session() as session:
             headers = {'Content-Type': content_type}
 
             try:
-                if size > 0:
+                if size <= 0:
+                    headers.update({'Content-Size': '0'})
+                    r = session.put(url, data='', headers=headers, timeout=5)
+                # for files under half a GB
+                elif size <= (10e8) / 2:
                     with open(path, 'rb') as f:
                         r = session.put(
                             url, data=f, headers=headers, timeout=5)
+                # # for chonky files, use a multipart upload
                 else:
-                    headers.update({'Content-Size': '0'})
-                    r = session.put(url, data='', headers=headers, timeout=5)
-
-                cls.validate_s3_response(r)
+                    # Chunks need to be at least 5MB or AWS throws an
+                    # EntityTooSmall error; we'll arbitrarily choose a
+                    # 15MB chunksize
+                    #
+                    # Note also that AWS limits the max number of chunkc
+                    # in a multipart upload to 10000, so this setting
+                    # currently enforces a hard limit on 150GB per file.
+                    #
+                    # We can dynamically assign a larger part size if needed,
+                    # but for the majority of use cases we should be fine
+                    # as-is
+                    part_minsize = int(15e6)
+                    dataset_id, _, version = dataset_version_id.partition(":")
+                    mpu_url = f'/datasets/{dataset_id}/versions/{version}/s3/preSignedUrls'
+
+                    api_client = http_client.API(
+                        api_url=config.CONFIG_HOST,
+                        api_key=self.api_key,
+                        ps_client_name=CLI_PS_CLIENT_NAME
+                    )
+
+                    mpu_create_res = api_client.post(
+                        url=mpu_url,
+                        json={
+                            'datasetId': dataset_id,
+                            'version': version,
+                            'calls': [{
+                                'method': 'createMultipartUpload',
+                                'params': {'Key': key}
+                            }]
+                        }
+                    )
+                    mpu_data = json.loads(mpu_create_res.text)[0]['url']
+
+                    parts = []
+                    with open(path, 'rb') as f:
+                        # we +2 the number of parts since we're doing floor
+                        # division, which will cut off any trailing part
+                        # less than the part_minsize, AND we want to 1-index
+                        # our range to match what AWS expects for part
+                        # numbers
+                        for part in range(1, (size // part_minsize) + 2):
```

Comment:
this'll also add an extra empty part if the upload is _exactly_ divisible by 500MB, which will probably cause an error from AWS due to it being too small. But also 🤷

Reply:
Really???? Use `ceil`...